internal/controller/config.go

```diff
@@ -0,0 +1,42 @@
+package controller
+
+import (
+	"fmt"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/spf13/viper"
+	corev1 "k8s.io/api/core/v1"
+)
+
+// GlobalConfig is the configuration for the controllers, it is
+// supposed to be populated from the configmap attached to the
+// controller manager.
+type GlobalConfig struct {
+	ClusterName     string                               `yaml:"clusterName" json:"clusterName"`
+	TimeInterval    uint                                 `yaml:"timeInterval" json:"timeInterval"`
+	RepoCredentials map[string]*corev1.SecretKeySelector `json:"repoCredentials"`
+}
```

Comment:
I think we should change the `RepoCredentials` a bit to be able to incorporate HTTP, SSH, and TLS - let me get back to you on that

Reply:
I'm thinking something like this:
```
type GitCredential struct {
    HTTPCredential *HTTPCredential 
    SSHCredential *SSHCredential
    TLS *TLS
}

type HTTPCredential struct {
    Username string
    Password *corev1.SecretKeySelector
}
type SSHCredential struct {
   Name string
   SSHKey *corev1.SecretKeySelector
}

type TLS struct {
	// +optional
	InsecureSkipVerify bool `json:"insecureSkipVerify,omitempty" protobuf:"bytes,1,opt,name=insecureSkipVerify"`
	// CACertSecret refers to the secret that contains the CA cert
	// +optional
	CACertSecret *corev1.SecretKeySelector `json:"caCertSecret,omitempty" protobuf:"bytes,2,opt,name=caCertSecret"`
	// CertSecret refers to the secret that contains the cert
	// +optional
	CertSecret *corev1.SecretKeySelector `json:"clientCertSecret,omitempty" protobuf:"bytes,3,opt,name=certSecret"`
	// KeySecret refers to the secret that contains the key
	// +optional
	KeySecret *corev1.SecretKeySelector `json:"clientKeySecret,omitempty" protobuf:"bytes,4,opt,name=keySecret"`
}
```

Reply:
What I'm thinking we'll put in the key for the `map[string]*GitCredential` is something like "github.intuit.com". You can imagine that somebody may specify a RepositoryPath like "https://github.intuit.com/myrepo" or "ssh://github.intuit.com/myrepo". We could specify both the SSHCredential there as well as the HTTPCredential, assuming that credentials are required for each of them. And the TLS config would apply to both of them.

Reply:
(sorry if we may need to iterate on what I have above at some point - this is what I'm thinking at the moment

Reply:
`*corev1.SecretKeySelector`  has an embedded struct  -```
type SecretKeySelector struct {
	// The name of the secret in the pod's namespace to select from.
	LocalObjectReference `json:",inline" protobuf:"bytes,1,opt,name=localObjectReference"`
	// The key of the secret to select from.  Must be a valid secret key.
	Key string `json:"key" protobuf:"bytes,2,opt,name=key"`
	// Specify whether the Secret or its key must be defined
	// +optional
	Optional *bool `json:"optional,omitempty" protobuf:"varint,3,opt,name=optional"`
}
```
because of which name is not able to get parsed by viper `Loadconfig` which is inside the `LocalObjectReference `struct

```
type LocalObjectReference struct {
	// Name of the referent.
	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
	// TODO: Add other useful fields. apiVersion, kind, uid?
	// +optional
	Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
}
```
We can use `corev1.SecretKeySelector` if we dnt require name key otherwise its limitation with viper -see this https://github.com/spf13/viper/issues/797
I can replicate corev1.SecretKeySelector and can add the `mapstructure:",squash"` that might solve the problem

See working issue -https://replit.com/join/qkzwkrlabn-shubhamdixit863

Reply:
Great research. 
> I can replicate corev1.SecretKeySelector and can add the `mapstructure:",squash"` that might solve the problem

That sounds good I think!

Reply:
Let's make sure we're keeping viper - have you looked at this [concurrency](https://github.com/numaproj-labs/numaplane/pull/80#issuecomment-1934681589) issue at all? Do you know what I mean?

Reply:
yeah i understood the point as the callback for reload can change the shared Globalconfig that might cause the race condition .I will initialize the mutex and and add a lock probably here . I think RW lock would be better in this scenario as we will have less frequent writes as compared to reads what do you think @juliev0